### PR TITLE
CVL-1442 Improve Prison API error handling

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/gov/GovUkApiClient.kt [hunk 1]
```diff
@@ -20,6 +20,6 @@ class GovUkApiClient(@Qualifier("govUkWebClient") val govUkApiClient: WebClient)
       .block()
 
     return response?.bankHolidayResult?.events?.map { it.date }
-      ?: throw IllegalStateException("Unexpected null response from API")
+      ?: error("Unexpected null response from API")
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/gov/GovUkApiClient.kt:23]: remove detekt error

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt [hunk 2]
```diff
@@ -84,7 +84,7 @@ class UpdateSentenceDateService(
     sentenceChanges: SentenceChanges,
   ) {
     updatedLicenceEntity.bookingId?.let {
-      prisonApiClient.hdcStatus(it).defaultIfEmpty(PrisonerHdcStatus(passed = false, approvalStatus = "UNKNOWN"))
+      prisonApiClient.getHdcStatus(it).defaultIfEmpty(PrisonerHdcStatus(passed = false, approvalStatus = "UNKNOWN"))
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/UpdateSentenceDateService.kt:87]: following java naming conventions

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt [hunk 3]
```diff
@@ -37,19 +36,6 @@ class PrisonApiClient(@Qualifier("oauthPrisonClient") val prisonerApiWebClient:
       .bodyValue(bookingIds)
       .retrieve()
       .bodyToMono(typeReference<List<PrisonerHdcStatus>>())
-      .onErrorResume { webClientErrorHandler(it) }
-      .block() ?: emptyList()
-  }
-
-  fun getOffenceHistories(bookingIds: List<Long>): List<PrisonerOffenceHistory> {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt:44]: not used any more

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt [hunk 4]
```diff
@@ -61,29 +47,18 @@ class PrisonApiClient(@Qualifier("oauthPrisonClient") val prisonerApiWebClient:
       .accept(MediaType.APPLICATION_JSON)
       .retrieve()
       .bodyToMono(typeReference<List<CourtEventOutcome>>())
-      .onErrorResume { webClientErrorHandler(it) }
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt:64]: This should never return an 404 - just an empty list

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/wiremock/PrisonApiMockServer.kt [hunk 5]
```diff
@@ -3,12 +3,13 @@ package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.integration.wiremo
...
 import com.github.tomakehurst.wiremock.client.WireMock.aResponse
 import com.github.tomakehurst.wiremock.client.WireMock.get
+import com.github.tomakehurst.wiremock.client.WireMock.post
 import com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo
 
 class PrisonApiMockServer : WireMockServer(8091) {
   fun stubGetHdcLatest(bookingId: Long = 12345, approvalStatus: String = "REJECTED", passed: Boolean = true) {
     stubFor(
-      get(urlEqualTo("/offender-sentences/booking/$bookingId/home-detention-curfews/latest")).willReturn(
+      get(urlEqualTo("/api/offender-sentences/booking/$bookingId/home-detention-curfews/latest")).willReturn(
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/wiremock/PrisonApiMockServer.kt:12]: This never worked-  was coercing 404s to nullable

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt [hunk 6]
```diff
@@ -61,29 +47,18 @@ class PrisonApiClient(@Qualifier("oauthPrisonClient") val prisonerApiWebClient:
...
       .block() ?: emptyList()
   }
 
-  private fun <API_RESPONSE_BODY_TYPE> webClientErrorHandler(exception: Throwable): Mono<API_RESPONSE_BODY_TYPE> =
+  private fun <API_RESPONSE_BODY_TYPE> coerce404ResponseToNull(exception: Throwable): Mono<API_RESPONSE_BODY_TYPE> =
     with(exception) {
-      if (this is WebClientResponseException) {
-        val uriPath = request?.uri?.path
-        when (statusCode) {
-          FORBIDDEN -> {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt:73]: This should be treat as an unexpected error and fail LOUDLY

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt [hunk 7]
```diff
@@ -61,29 +47,18 @@ class PrisonApiClient(@Qualifier("oauthPrisonClient") val prisonerApiWebClient:
...
-            log.error("Client token does not have correct role to call prisoner-api $uriPath")
-          }
-
-          NOT_FOUND -> {
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt:77]: treating this as an acceptable behaviour in a generic error handler is unsafe behaviour.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt [hunk 8]
```diff
@@ -61,29 +47,18 @@ class PrisonApiClient(@Qualifier("oauthPrisonClient") val prisonerApiWebClient:
...
-            log.info("No resource found when calling prisoner-api $uriPath")
-          }
-
-          else -> {
-            log.error("Failed to call prisoner-api $uriPath [statusCode: $statusCode, body: ${this.responseBodyAsString}]")
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt:82]: This wasn't logging an exception as a separate parameter meaning it wouldn't be recorded as an exception in app insights

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt [hunk 9]
```diff
@@ -61,29 +47,18 @@ class PrisonApiClient(@Qualifier("oauthPrisonClient") val prisonerApiWebClient:
       .accept(MediaType.APPLICATION_JSON)
       .retrieve()
       .bodyToMono(typeReference<List<CourtEventOutcome>>())
-      .onErrorResume { webClientErrorHandler(it) }
       .block() ?: emptyList()
   }
 
-  private fun <API_RESPONSE_BODY_TYPE> webClientErrorHandler(exception: Throwable): Mono<API_RESPONSE_BODY_TYPE> =
+  private fun <API_RESPONSE_BODY_TYPE> coerce404ResponseToNull(exception: Throwable): Mono<API_RESPONSE_BODY_TYPE> =
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/prison/PrisonApiClient.kt:53]: remodelled this to only handle situations where you would explicitly want to treat a 404 response as a valid case that represents the response being optional

**Review by nishanthmahasamudram — APPROVED** [review]
LGTM

**Pull request merged**